File: android/app/src/main/kotlin/org/fgsec/halogen/features/settings/UsernameEditView.kt
package org.fgsec.halogen.features.settings

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.input.KeyboardCapitalization
import androidx.compose.ui.unit.dp
import kotlinx.coroutines.launch
import org.fgsec.halogen.components.ToastCenter
import org.fgsec.halogen.core.FriendlyError
import org.fgsec.halogen.core.HalogenCore

/// Edit the signed-in user's username (`PUT /users/{id}`, self-or-admin).
/// The stored session updates so the change survives relaunch.
@Composable
fun UsernameEditView(core: HalogenCore, onBack: () -> Unit) {
    var username by remember { mutableStateOf(core.account?.username ?: "") }
    var error by remember { mutableStateOf<String?>(null) }
    var saving by remember { mutableStateOf(false) }
    val scope = rememberCoroutineScope()
    val trimmed = username.trim()

    suspend fun save() {
        val account = core.account ?: return
        saving = true
        try {
            core.updateUsername(userId = account.userId, username = trimmed)
            core.renameActiveAccount(trimmed)
            ToastCenter.success("Username updated")
            onBack()
        } catch (e: Exception) {
            error = FriendlyError.message(e)
        } finally {
            saving = false
        }
    }

    SettingsScaffold(title = "Edit username", onBack = onBack) { padding ->
        Column(
            Modifier.fillMaxSize().padding(padding).verticalScroll(rememberScrollState()),
        ) {
            SettingsGroup {
                SettingsTextField(
                    value = username,
                    onValueChange = { username = it },
                    label = "Username",
                    capitalization = KeyboardCapitalization.None,
                )
            }
            error?.let {
                FootnoteText(
                    it,
                    color = MaterialTheme.colorScheme.error,
                    modifier = Modifier.padding(horizontal = 32.dp),
                )
            }
            Box(Modifier.padding(16.dp)) {
                // Web rules: 3-64 chars; compare the TRIMMED value (a padded
                // copy of the same name used to pass the unchanged guard and
                // PUT an identical name); offline-gated (online-only endpoint).
                BusyButton(
                    title = "Save",
                    busy = saving,
                    enabled = trimmed.length >= 3 && trimmed.length <= 64 &&
                        trimmed != core.account?.username && !core.isOffline,
                ) {
                    scope.launch { save() }
                }
            }
            if (core.isOffline) {
                FootnoteText(
                    "You're offline — reconnect to rename the account.",
                    modifier = Modifier.padding(horizontal = 32.dp),
                )
            }
        }
    }
}

File: ios/Halogen/Features/Settings/UsernameEditView.swift
import SwiftUI

/// Edit the signed-in user's username (`PUT /users/{id}`, self-or-admin).
/// The stored session updates so the change survives relaunch.
struct UsernameEditView: View {
    let core: HalogenCore

    @Environment(\.dismiss) private var dismiss
    @State private var username: String
    @State private var error: String?
    @State private var saving = false

    init(core: HalogenCore) {
        self.core = core
        _username = State(initialValue: core.account?.username ?? "")
    }

    var body: some View {
        Form {
            TextField("Username", text: $username)
                .autocorrectionDisabled()
                .textInputAutocapitalization(.never)
            if let error {
                Text(error).font(.footnote).foregroundStyle(.red)
            }
            Button {
                Task { await save() }
            } label: {
                if saving {
                    ProgressView().frame(maxWidth: .infinity)
                } else {
                    Text("Save").frame(maxWidth: .infinity)
                }
            }
            // Web rules: 3-64 chars; compare the TRIMMED value (a padded
            // copy of the same name used to pass the unchanged guard and PUT
            // an identical name); offline-gated (online-only endpoint).
            .disabled(
                saving || trimmedUsername.count < 3 || trimmedUsername.count > 64
                    || trimmedUsername == core.account?.username || core.isOffline)
            if core.isOffline {
                Text("You're offline — reconnect to rename the account.")
                    .font(.footnote).foregroundStyle(.secondary)
            }
        }
        .navigationTitle("Edit username")
        .navigationBarTitleDisplayMode(.inline)
    }

    private var trimmedUsername: String {
        username.trimmingCharacters(in: .whitespaces)
    }

    private func save() async {
        guard let account = core.account else { return }
        saving = true
        defer { saving = false }
        do {
            try await core.updateUsername(
                userId: account.userId,
                username: username.trimmingCharacters(in: .whitespaces))
            core.renameActiveAccount(to: username.trimmingCharacters(in: .whitespaces))
            ToastCenter.shared.success("Username updated")
            dismiss()
        } catch {
            self.error = FriendlyError.message(error)
        }
    }
}
